sound/soc/sof/core.c

```diff
@@ -365,6 +373,9 @@ int snd_sof_device_remove(struct device *dev)
 	if (IS_ENABLED(CONFIG_SND_SOC_SOF_PROBE_WORK_QUEUE))
 		cancel_work_sync(&sdev->probe_work);
 
+	/* Unregister any registered client device */
+	sof_unregister_clients(sdev);
+
```

Comment:
The `snd_sof_free_debug()` later would invoke the `debugfs_remove_recursive()` on the debugfs root of SOF and instead of doing a check in every client's remove I thought it might be better to unregister them while we still have the debugfs directory.
The clients are mainly IPC clients and they might need to do IPC as part of their remove, so keeping it ahead of `snd_sof_ipc_free()` sounded sensible also.

Reply:
makes sense @ujfalusi, you should add a comment to make mention of this point I missed :-)

Also could we move the snd_sof_machine_unregister() before line 379, so that clients and machine are removed before the core releasing resources?